src/hv_aic_events.c

```diff
@@ -0,0 +1,166 @@
+                break;
+            case 1:
+                write16(addr, val[0]);
+                break;
+            case 2:
+                write32(addr, val[0]);
+                break;
+            case 3:
+                write64(addr, val[0]);
+                break;
+            case 4:
+                write64(addr, val[0]);
+                write64(addr + 8, val[1]);
+                break;
+            default:
+                dprintf("HV: unsupported width %ld\n", width);
+                return false;
+        }
+    } else {
+        switch (width) {
+            case 0:
+                val[0] = read8(addr);
+                break;
+            case 1:
+                val[0] = read16(addr);
+                break;
+            case 2:
+                val[0] = read32(addr);
+                break;
+            case 3:
+                val[0] = read64(addr);
+                break;
+            case 4:
+                val[0] = read64(addr);
+                val[1] = read64(addr + 8);
+                break;
+            default:
+                dprintf("HV: unsupported width %ld\n", width);
+                return false;
+        }
+    }
+
+    if (addr != (aic_base + AIC_EVENT) || write || width != 2) {
+        return true;
+    }
+
+    u16 type = (*val & AIC_EVENT_TYPE) >> 16;
+    u16 num = *val & AIC_EVENT_NUM;
+
+    switch (type) {
+        case AIC_EVENT_TYPE_HW:
+            if (check_hw_num(num)) {
+                emit_irqtrace(type, num);
+            }
+            break;
+        default:
+            // ignore
+            break;
+    }
+
+    return true;
+}
+
+void hv_trace_aic_events(u32 type, u32 num, u32 active)
+{
+    switch (type) {
+        case AIC_EVENT_TYPE_HW:
+            if (num >= 1024) {
+                printf("hv: invalid IRQ number %u\n", num);
+                return;
+            }
+            dprintf("%s tracing of irq %us\n", active ? "activating" : "deactivating", num);
+            if (active)
+                set_hw_num(num);
+            else
+                clear_hw_num(num);
+            break;
+        case AIC_EVENT_TYPE_IPI:
+            printf("hv: Ignoring IPI AIC event 0x%04x\n", num);
+            return;
+        default:
+            printf("hv: Ignoring unkown AIC event type: 0x%02x num: %u\n", type, num);
+            return;
+    }
+
+    if (!aic_base) {
+        static const char path[] = "/arm-io/aic";
+        int adt_path[8];
+
+        int node = adt_path_offset_trace(adt, path, adt_path);
+        if (node < 0) {
+            printf("hv: Error getting %s node\n", path);
+            return;
+        }
+
+        if (adt_get_reg(adt, adt_path, "reg", 0, &aic_base, NULL) < 0) {
+            printf("hv: Error getting AIC base address.\n");
+            return;
+        }
+    }
+
+    static int hooked = 0;
+
+    if (aic_base && !hooked) {
+        hv_map_hook(aic_base, trace_aic_event, 0x4000);
+        hooked = 1;
+    }
```

Comment:
Maybe complain if we can't find the AIC, possibly actually return an error code to the proxy?

Reply:
all done I hope